src/ray/gcs/gcs_server/gcs_server.cc

```diff
@@ -287,5 +288,13 @@ std::unique_ptr<GcsWorkerManager> GcsServer::InitGcsWorkerManager() {
       new GcsWorkerManager(gcs_table_storage_, gcs_pub_sub_));
 }
 
+void GcsServer::CollectStats() {
+  gcs_actor_manager_->CollectStats();
```

Comment:
Is this an infinite timer loop?
It should be stopped in destructor.

Reply:
This uses the main_service, so I think this will be destructed if main_service is destructed? Btw, I noticed that we didn't call `CollectStats` at all in gcs_server.cc. I will fix that.

Reply:
> This uses the main_service, so I think this will be destructed if main_service is destructed? Btw, I noticed that we didn't call `CollectStats` at all in gcs_server.cc. I will fix that.

We'd better make sure any stats collecting can not block main service of gcs. Actually it's unsafe to invoke the function `CollectStats` out of gcs server instance scope even if main service is still alive.

Reply:
Hmm I don't think I understand what you meant. Do you suggest to use a separate thread? It is tricky because gcs server is not meant to be multi threads now. Also, can you elaborate what this means? `it's unsafe to invoke the function CollectStats out of gcs server instance scope even if main service is still alive.`?

Reply:
In some tests, I mean collect stats will keep running timely while gcs server has been destroyed so `this` pointer is invalid when it's in out of gcs server instance lifecycle.

Reply:
Hmm... That's the same for everything else that uses main_service right? When destructor is invoked, if there are remaining tasks in the main_service, that'll be the same although we don't have this method.

Reply:
I don't think separate thread could be better in any way. This is invoked every 2~3 seconds, and there's no way this overhead can impact gcs server (If so, we should redesign the gcs server). Using a separate thread for this require to make gcs server thread safe, which makes things 100 times more complicated.

Reply:
A seperate thread calls read-only method `marked in constant`, so no fields in gcs server will be modified.

Reply:
BTW, the new thread or service is not in gcs server, I mean, we do not redesign gcs server or task dispacher pipeline.
A thread in stats module can invoke collecting function periodically and this collecting function is registered from different metrics by `Stats::RegisterMetricWithLambda(Metric, function)`.

Reply:
In stats timer generating function,
```cpp
for (pair<metric, function> item in registered_list) {
    item.first.record(function());
}
```
We can unify collecting operation of all components in this way without defining colleting_stats function anywhere.

Reply:
> A seperate thread calls read-only method marked in constant, so no fields in gcs server will be modified.

It is still possible fields that we need to read is accessed and modified by other threads.

Also, it'll be great how that approach. can be better than the current approach. I don't think I see benefits from your suggested implementation (it will only make it complicated because multiple threads will need to access fields in all classes).

Reply:
To simplify problem, only one thread is used in gcs. But it's a different scenario in raylet and coreworker. For example, we may create a new thread for caculating cpu/memory usage or getting from other system command because it doesn't belong to any of them.

Reply:
Those will be collected by dashboard agent since they are already collected there. We will only collect component-specific metrics from components. Also, coreworker and raylet collect these things in the same way as current implementation.

Reply:
Anyway, if you'd like to collect default set of "system" metrics and want to add some logic within stats module, that's fine. But extending that pattern to collect component stats require it to be thread-safe, and I don't think it is worth doing that now. In my opinion, there's a tiny or nearly no benefit of doing this at this point unless we'd like to collect metrics with much less frequency. So I will just keep the implementation. If you guys have some ways to do this and contribute, I'd love to review the PR.